Sources/ClientRuntime/Util/Async.swift

```diff
@@ -0,0 +1,53 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Dispatch
+
+private class BoxedThrowingResult<ResultType> {
+    var result: Result<ResultType, Error>?
+}
+
+private class BoxedResult<ResultType> {
+    var result: ResultType?
+}
+
+/// Unsafely awaits a throwing async function from a synchronous context.
+@available(*, deprecated, message: """
+    This is a temporary workaround for the lack of async support.
+    It will be migrated to structured concurrency in the future
+""")
+public func _unsafeThrowingWait<ResultType>(_ f: @escaping () async throws -> ResultType) throws -> ResultType {
+    let box = BoxedThrowingResult<ResultType>()
+    let semaphore = DispatchSemaphore(value: 0)
+    Task {
+        do {
+            let val = try await f()
+            box.result = .success(val)
+        } catch {
+            box.result = .failure(error)
+        }
+        semaphore.signal()
+    }
+    semaphore.wait()
+    return try box.result!.get()
```

Comment:
Can we just use Result here and remove `BoxedThrowingResult`? IMO its simpler 

```suggestion
    var result: Result<ResultType, Error>? = nil
    let semaphore = DispatchSemaphore(value: 0)
    Task {
        do {
            let val = try await f()
            result = .success(val)
        } catch {
            result = .failure(error)
        }
        semaphore.signal()
    }
    semaphore.wait()
    return try result!.get()
```

Reply:
Thats because we need a type that can be manipulated outside concurrently. Result is enum which is not possible to be altered if declared outside the concurrent block ie Task.

Reply:
Ahh thats right